### PR TITLE
Add FXIOS-9201 ⁃ Enhanced Tracking Protection certificates details screen

### PR DESCRIPTION
**Pull request by dicarobinho**
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-9201)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/20369)

## :bulb: Description
Added Tracking Protection Certificates screen

## :pencil: Checklist
You have to check all boxes before merging
- [x] Filled in the above information (tickets numbers and description of your work)
- [x] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [ ] Wrote unit tests and/or ensured the tests suite is passing
- [ ] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [ ] If needed, I updated documentation / comments for complex code and public methods
- [ ] If needed, added a backport comment (example `@Mergifyio backport release/v120`)

### CODE REVIEW

**Comment by ih-codes** [comment]
Sorry I didn't get a chance to review this on Friday @dicarobinho. Would you be able to provide instructions for how I can access this screen to test in the app? Thanks!

File: firefox-ios/Client/Frontend/TrackingProtection/CertificatesViewController.swift [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+import X509
+import SwiftASN1
+
+struct CertificateKeys {
+    static let commonName = "CN"
+    static let country = "C"
+    static let organization = "O"
+}
+
+struct CertificatesUX {
```
Comment (ih-codes) [inline firefox-ios/Client/Frontend/TrackingProtection/CertificatesViewController.swift:21]: I think we'd like to embed the `CertificateKeys` and `CertificatesUX` inside the `CertificatesViewController`. I double checked in slack, and that sounds like the style people prefer to move toward for UX constants (example: the generalized `UX` struct in `CreditCardBottomSheetViewController`).

File: firefox-ios/Client/Frontend/TrackingProtection/CertificatesViewController.swift [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+    private let titleLabel: UILabel = .build { label in
+        label.font = FXFontStyles.Bold.title1.scaledFont()
+        label.text = .Menu.EnhancedTrackingProtection.certificatesTitle
+    }
+
+    private let headerStackView: UIStackView = .build { stack in
+        stack.axis = .horizontal
+        stack.distribution = .fillEqually
+        stack.spacing = CertificatesUX.headerStackViewSpacing
+        stack.translatesAutoresizingMaskIntoConstraints = false
```
Comment (ih-codes) [inline firefox-ios/Client/Frontend/TrackingProtection/CertificatesViewController.swift:50]: One of the things the `.build` helper does is set `translatesAutoresizingMaskIntoConstraints` to `false` so that's probably unnecessary in these view/constraint setups (also lines 54).

Also, if you use the `.build` setup for the `certificatesTableView`, you could put some of the view setup there and save `setupCertificatesTableView` just for constraints. And then you also wouldn't need line 154.

File: firefox-ios/Client/Frontend/TrackingProtection/TrackingProtectionModel.swift [hunk 3]
```diff
@@ -136,6 +136,13 @@ class TrackingProtectionModel {
         }
     }
 
+    func getCertificatesViewModel() -> CertificatesViewModel {
```
Comment (ih-codes) [inline firefox-ios/Client/Frontend/TrackingProtection/TrackingProtectionModel.swift:139]: Do we ever need multiple instances of the CertificatesViewModel? If no, this can simply be a lazy var `certificatesViewModel`.

File: firefox-ios/Client/Frontend/TrackingProtection/TrackingProtectionViewController.swift [hunk 4]
```diff
@@ -334,6 +334,11 @@ class TrackingProtectionViewController: UIViewController, Themeable, Notifiable,
         updateViewDetails()
         updateProtectionViewStatus()
         applyTheme()
+        getCertificates(for: viewModel.url) { certificates in
```
Comment (ih-codes) [inline firefox-ios/Client/Frontend/TrackingProtection/TrackingProtectionViewController.swift:337]: We probably want a `weak self` here if this call could run long to avoid retaining the viewController.

File: firefox-ios/Client/Frontend/TrackingProtection/CertificatesCell.swift [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+import Foundation
+import Common
+
+final class CertificatesCell: UITableViewCell, ReusableCell, ThemeApplicable {
+    var sectionLabel: UILabel = .build { label in
+        label.font = FXFontStyles.Bold.headline.scaledFont()
+        label.textAlignment = .right
+        label.numberOfLines = 0
+        label.lineBreakMode = .byWordWrapping
+        label.translatesAutoresizingMaskIntoConstraints = false
```
Comment (ih-codes) [inline firefox-ios/Client/Frontend/TrackingProtection/CertificatesCell.swift:14]: (See my other comment regarding the `.build` method). Same for line 84 in this file.

File: firefox-ios/Client/Frontend/TrackingProtection/CertificatesHandler.swift [hunk 6]
```diff
@@ -20,17 +19,67 @@ class CertificatesHandler {
...
+    init(completion: @escaping ([Certificate]?) -> Void) {
+        self.completion = completion
+    }
+
+    func urlSession(_ session: URLSession, didReceive
+                    challenge: URLAuthenticationChallenge,
+                    completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void) {
+        if let serverTrust = challenge.protectionSpace.serverTrust {
+            let certificatesHandler = CertificatesHandler(serverTrust: serverTrust)
+            certificatesHandler.handleCertificates { certificates in
```
Comment (ih-codes) [inline firefox-ios/Client/Frontend/TrackingProtection/CertificatesHandler.swift:74]: It's a bit misleading to need a completion handler for `handleCertificates` since no async work is being done. Maybe this method should simply return `[Certificate]`, which can then be passed to the `CertificateDelegate`'s `completion`.

That way it's 100% clear from looking at this method that all code paths definitely call's the delegate's `completion`.

File: firefox-ios/Client/Frontend/TrackingProtection/CertificatesViewController.swift [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+    static let titleLabelMargin = 8.0
+    static let titleLabelTopMargin = 20.0
+    static let headerStackViewMargin = 8.0
+    static let headerStackViewTopMargin = 20.0
+    static let tableViewSpacerTopMargin = 20.0
+    static let tableViewSpacerHeight = 1.0
+    static let tableViewTopMargin = 20.0
+}
+
+typealias CertificateItems = [(key: String, value: [String])]
```
Comment (ih-codes) [inline firefox-ios/Client/Frontend/TrackingProtection/CertificatesViewController.swift:37]: As far as I can tell, `value` is never actually an array. So `value` can just be `String` (involves some refactoring). That'll also make `tableView:cellForRowAt` a bit easier to read.

**Review by ih-codes — COMMENTED** [review]
Okay, I've added a few comments. Mostly nitpicks (sorry!).

The `CertificatesViewModel` has some complicated processing logic. I think it would benefit from adding some unit tests.

File: firefox-ios/Client/Frontend/TrackingProtection/TrackingProtectionModel.swift (hunk 3 above)
Comment (cyndichin) [reply firefox-ios/Client/Frontend/TrackingProtection/TrackingProtectionModel.swift:139]: also instead of view models, we should use redux. I would like to see a TODO ticket if we want to do this in a separate PR. This logic seems like it should live in a certificate manager that lives in the middleware

File: firefox-ios/Client/Frontend/TrackingProtection/TrackingProtectionViewController.swift (hunk 4 above)
Comment (cyndichin) [reply firefox-ios/Client/Frontend/TrackingProtection/TrackingProtectionViewController.swift:337]: why does the tracking protection VC need to know about this? to better understand, could this be done in the initialization of the certificate view?

**Review by cyndichin — COMMENTED** [review]
@dicarobinho I haven't reviewed thoroughly yet, will give another look once @ih-codes comments are addressed. However, we will be using the certificate screen for the error pages as well. Is this view reusable in that case? 

Please provide screenshots as well, it will help review. Thank you!

File: firefox-ios/Client/Frontend/TrackingProtection/CertificatesViewController.swift (hunk 1 above)
Comment (dicarobinho) [reply firefox-ios/Client/Frontend/TrackingProtection/CertificatesViewController.swift:21]: Ok, I can do that but that will cause some errors in my cell because I use also there some UX values. May I separate them? UX for cell and UX for Controller?

File: firefox-ios/Client/Frontend/TrackingProtection/TrackingProtectionModel.swift (hunk 3 above)
Comment (dicarobinho) [reply firefox-ios/Client/Frontend/TrackingProtection/TrackingProtectionModel.swift:139]: @ih-codes yes, as I know this may used in future.
@cyndichin I am going to add a TODO

File: firefox-ios/Client/Frontend/TrackingProtection/TrackingProtectionViewController.swift (hunk 4 above)
Comment (dicarobinho) [reply firefox-ios/Client/Frontend/TrackingProtection/TrackingProtectionViewController.swift:337]: @ih-codes yes, [weak self] may be better there
@cyndichin may you be more exactly, please?

File: firefox-ios/Client/Frontend/TrackingProtection/CertificatesViewController.swift (hunk 7 above)
Comment (dicarobinho) [reply firefox-ios/Client/Frontend/TrackingProtection/CertificatesViewController.swift:37]: If you take a look in getDNSNames method, from viewModel, there is a case where we need a value as an array because for DNSName key we have a list of values.

**Comment by dicarobinho** [comment]
> @dicarobinho I haven't reviewed thoroughly yet, will give another look once @ih-codes comments are addressed. However, we will be using the certificate screen for the error pages as well. Is this view reusable in that case?
> 
> Please provide screenshots as well, it will help review. Thank you!

Screen is not done yet, so screenshots may not be useful at this step. (I have another TODO done for which I will create a PR after this one will be merged.)

File: firefox-ios/Client/Frontend/TrackingProtection/TrackingProtectionViewController.swift (hunk 4 above)
Comment (cyndichin) [reply firefox-ios/Client/Frontend/TrackingProtection/TrackingProtectionViewController.swift]: I would like to understand why we would like to initialize `getCertificates` in the `TrackingViewController`. I could be misinterpreting this, so wanted to understand the reasoning behind adding the code here.

**Comment by cyndichin** [comment]
> > @dicarobinho I haven't reviewed thoroughly yet, will give another look once @ih-codes comments are addressed. However, we will be using the certificate screen for the error pages as well. Is this view reusable in that case?
> > Please provide screenshots as well, it will help review. Thank you!
> 
> Screen is not done yet, so screenshots may not be useful at this step. (I have another TODO done for which I will create a PR after this one will be merged.)

Even if the screen is not done, it'll be helpful to add it to represent what code is being merged from a user perspective. At least from my side, it'll help me provide a quicker review to visualize. :)

**Comment by dicarobinho** [comment]
> > > @dicarobinho I haven't reviewed thoroughly yet, will give another look once @ih-codes comments are addressed. However, we will be using the certificate screen for the error pages as well. Is this view reusable in that case?
> > > Please provide screenshots as well, it will help review. Thank you!
> > 
> > 
> > Screen is not done yet, so screenshots may not be useful at this step. (I have another TODO done for which I will create a PR after this one will be merged.)
> 
> Even if the screen is not done, it'll be helpful to add it to represent what code is being merged from a user perspective. At least from my side, it'll help me provide a quicker review to visualize. :)

Simulator Screenshot - iPhone 15 Pro Max - 2024-08-27 at 16 37 34

File: firefox-ios/Client/Frontend/TrackingProtection/CertificatesViewController.swift (hunk 1 above)
Comment (ih-codes) [reply firefox-ios/Client/Frontend/TrackingProtection/CertificatesViewController.swift:21]: Yes, I think that makes sense, since none of the constants are used by both the cell and the controller (as far as I can tell).

So you can add a UX struct inside your cell (with just the constants needed in the cell), and a UX struct inside the view controller (with just the constants needed in the view controller). 👍

File: firefox-ios/Client.xcodeproj/project.pbxproj [hunk 8]
```diff
@@ -16,11 +16,18 @@
...
 		0A49784A2C53E63200B1E82A /* TrackingProtectionViewController.swift in Sources */ = {isa = PBXBuildFile; fileRef = 0A4978492C53E63200B1E82A /* TrackingProtectionViewController.swift */; };
+		0A7693612C7DD19600103A6D /* CertificatesModelTests.swift in Sources */ = {isa = PBXBuildFile; fileRef = 0A7693602C7DD19500103A6D /* CertificatesModelTests.swift */; };
 		0AC659272BF35854005C614A /* FxAWebViewModelTests.swift in Sources */ = {isa = PBXBuildFile; fileRef = 0AC659262BF35854005C614A /* FxAWebViewModelTests.swift */; };
 		0AC659292BF493CE005C614A /* MockFxAWebViewModel.swift in Sources */ = {isa = PBXBuildFile; fileRef = 0AC659282BF493CE005C614A /* MockFxAWebViewModel.swift */; };
 		0AD3EEAC2C2485A7001044E5 /* ThemedCenteredTableViewCell.swift in Sources */ = {isa = PBXBuildFile; fileRef = 0AD3EEAB2C2485A7001044E5 /* ThemedCenteredTableViewCell.swift */; };
+		0AFF7F5E2C74DCA300265214 /* (null) in Sources */ = {isa = PBXBuildFile; };
+		0AFF7F602C75D3D300265214 /* (null) in Sources */ = {isa = PBXBuildFile; };
+		0AFF7F622C77432900265214 /* (null) in Sources */ = {isa = PBXBuildFile; };
```
Comment (ih-codes) [inline firefox-ios/Client.xcodeproj/project.pbxproj:25]: Hmm, these 3 lines look a bit odd. 🤔 

Each of these 3 keys (e.g. `0AFF7F622C77432900265214`) also have another `(null)` line lower in this file too (e.g. line 15090).

You may need to manually delete these lines (and the corresponding ones below), as I suspect something strange happened and they shouldn't be in the project file.

File: firefox-ios/Client/Frontend/TrackingProtection/CertificatesHandler.swift [hunk 9]
```diff
@@ -20,17 +19,64 @@ class CertificatesHandler {
     }
 
     /// Extracts and handles the certificate chain.
-    func handleCertificates() {
+    /// - Parameters:
+    ///   - completion: A completion block that provides the extracted certificates.
+    func handleCertificates() -> [Certificate] {
```
Comment (ih-codes) [inline firefox-ios/Client/Frontend/TrackingProtection/CertificatesHandler.swift:24]: Very happy with this change, thanks!!

File: firefox-ios/Client/Frontend/TrackingProtection/CertificatesViewController.swift [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+            return UITableViewCell()
+        }
+
+        guard !viewModel.certificates.isEmpty else {
+            return cell
+        }
+
+        let certificate = viewModel.certificates[viewModel.selectedCertificateIndex]
+
+        switch indexPath.row {
```
Comment (ih-codes) [inline firefox-ios/Client/Frontend/TrackingProtection/CertificatesViewController.swift:198]: This is super close!

One very small nitpick to make things look cleaner:

```
switch CertificatesItemType(rawValue: indexPath.row) {
        case .subjectName:
        ...
        case .issuerName:
        ...
        ...
        ...
        case .none: // Deals with the nil case. 
            break
}
```
I also suggested `.none` for the nil case. `default` is ok but the compiler will not give you a warning if, in the future, someone adds a new `CertificatesItemType` case that isn't handled here.

FYI if you're curious, two examples of this pattern are in `LegacyInactiveTabCell` line 146 and `AddCredentialViewController` line 156.

File: firefox-ios/Client/Frontend/TrackingProtection/CertificatesViewController.swift [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+    static let headerStackViewTopMargin = 20.0
+    static let tableViewSpacerTopMargin = 20.0
+    static let tableViewSpacerHeight = 1.0
+    static let tableViewTopMargin = 20.0
+}
+
+typealias CertificateItems = [(key: String, value: String)]
+
+class CertificatesViewController: UIViewController, Themeable, UITableViewDelegate, UITableViewDataSource {
+    private enum CertificatesItemType: Int {
```
Comment (ih-codes) [inline firefox-ios/Client/Frontend/TrackingProtection/CertificatesViewController.swift:40]: Thanks for adding this!

Small hint:
```
    private enum CertificatesItemType: Int {
        case subjectName = 0
        case issuerName
        case validity
        case subjectAltName
    }
```

Swift will automatically give the enums ascending number values. :)

File: firefox-ios/firefox-ios-tests/Tests/ClientTests/TrackingProtectionTests/CertificatesModelTests.swift [hunk 12]
```diff
@@ -0,0 +1,47 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/
+
+import Foundation
+import XCTest
+import X509
+@testable import Client
+
+final class CertificatesModelTests: XCTestCase {
```
Comment (ih-codes) [inline firefox-ios/firefox-ios-tests/Tests/ClientTests/TrackingProtectionTests/CertificatesModelTests.swift:10]: Can you please rename the file and class to `CertificatesViewModelTests`? It will be easier to find the tests for `CertificatesViewModel` in the future! ("View" is missing)

Great tests by the way, thanks!

File: firefox-ios/firefox-ios-tests/Tests/ClientTests/TrackingProtectionTests/CertificatesModelTests.swift [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+        XCTAssertEqual(result["CN"], "www.google.com")
+        XCTAssertEqual(result["O"], "")
+        XCTAssertEqual(result["C"], "US")
+    }
+
+    func testGetDNSNamesList() {
+        let input = #"DNSName("www.google.com"), DNSName("*www.google.com")"#
+        let result = viewModel.getDNSNamesList(from: input)
+        XCTAssertEqual(result, ["www.google.com", "*www.google.com"])
+    }
```
Comment (ih-codes) [inline firefox-ios/firefox-ios-tests/Tests/ClientTests/TrackingProtectionTests/CertificatesModelTests.swift:46]: Should we maybe add tests where you pass an empty string as input to your new functions?

e.g. `let result = viewModel.getCertificateValues(from: "")`
e.g. `let result = viewModel.getDNSNamesList(from:"")`

I would like to know how your code handles bad input.

You might also want to think about what happens if an empty `[]` or `[:]` is returned in CertificatesViewController, for example. (I'll leave a comment in `setupCertificatesHeaderView`.)

File: firefox-ios/Client/Frontend/TrackingProtection/CertificatesViewController.swift [hunk 14]
```diff
@@ -0,0 +1,280 @@
...
+                            (.Menu.EnhancedTrackingProtection.certificateValidityNotBefore,
+                                certificate.notValidBefore.toRFC822String()),
+                            (.Menu.EnhancedTrackingProtection.certificateValidityNotAfter,
+                                certificate.notValidAfter.toRFC822String())
+                           ])
+
+        case CertificatesItemType.subjectAltName.rawValue:
+            cell.configure(theme: currentTheme(),
+                           sectionTitle: .Menu.EnhancedTrackingProtection.certificateSubjectAltNames,
+                           items: viewModel.getDNSNames(for: certificate))
```
Comment (ih-codes) [inline firefox-ios/Client/Frontend/TrackingProtection/CertificatesViewController.swift:231]: Similar comment to line 131.

If `getDNSNames` returns an empty array `[]`, what should happen? I think this probably just makes a cell with only a sectionLabel, which is probably fine since this is unlikely to happen. But might be worth double checking!

File: firefox-ios/Client/Frontend/TrackingProtection/CertificatesViewModel.swift [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+    var selectedCertificateIndex: Int = 0
+
+    init(topLevelDomain: String, title: String, URL: String, certificates: [Certificate]) {
+        self.topLevelDomain = topLevelDomain
+        self.title = title
+        self.URL = URL
+        self.certificates = certificates
+    }
+
+    func getCertificateValues(from data: String) -> [String: String] {
```
Comment (ih-codes) [inline firefox-ios/Client/Frontend/TrackingProtection/CertificatesViewModel.swift:23]: Sorry I missed this before. Maybe you can rename the input `from input: String` like you have with `getDNSNamesList`? 

It is a tiny bit confusing to call a string "data" since Swift also has a `Data` type.

**Review by ih-codes — COMMENTED** [review]
Thanks @dicarobinho for all your changes! I appreciate you adding the unit tests as well. I have another round of feedback after looking at your update.

**Comment by ih-codes** [comment]
> > Sorry I didn't get a chance to review this on Friday @dicarobinho. Would you be able to provide instructions for how I can access this screen to test in the app? Thanks!
> 
> No, at this step is not possible without doing some changes locally.

I don't mind making a few changes in my code so I can test the screen. But if it is complicated, then don't worry about it. The screenshot you posted helps! 👍 I imagine more detailed testing can be done once the screen is finished.

File: firefox-ios/Client/Frontend/TrackingProtection/TrackingProtectionViewController.swift (hunk 4 above)
Comment (dicarobinho) [reply firefox-ios/Client/Frontend/TrackingProtection/TrackingProtectionViewController.swift]: @cyndichin just want to have the certificates loaded for the Certificates screen also for the details screen. Don't want to load them, for two separate times.

File: firefox-ios/Client.xcodeproj/project.pbxproj (hunk 8 above)
Comment (dicarobinho) [reply firefox-ios/Client.xcodeproj/project.pbxproj:25]: Don't even know how I achieved that. Thank you.

File: firefox-ios/Client/Frontend/TrackingProtection/CertificatesViewController.swift [hunk 16]
```diff
@@ -0,0 +1,280 @@
...
+            titleLabel.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -CertificatesUX.titleLabelMargin),
+            titleLabel.topAnchor.constraint(equalTo: view.topAnchor, constant: CertificatesUX.titleLabelTopMargin)
+        ])
+    }
+
+    private func setupCertificatesHeaderView() {
+        // TODO: FXIOS-9834 Add tab indicator for table view tabs and hide/unhide it on selection
+        for (index, certificate) in viewModel.certificates.enumerated() {
+            let button: UIButton = .build { [weak self] button in
+                button.setTitle(self?.viewModel.getCertificateValues(from: "\(certificate.subject)")[CertificateKeys.commonName], for: .normal)
```
[the remark this answers is not in the document]
Comment (dicarobinho) [reply firefox-ios/Client/Frontend/TrackingProtection/CertificatesViewController.swift:131]: Good catch. Let's skip (not add it) the button, in that case.

File: firefox-ios/Client/Frontend/TrackingProtection/CertificatesViewController.swift (hunk 14 above)
Comment (dicarobinho) [reply firefox-ios/Client/Frontend/TrackingProtection/CertificatesViewController.swift:231]: For websites, without DNSNames, subjectAltName cell will not have values. (only the section label).

File: firefox-ios/Client/Frontend/TrackingProtection/TrackingProtectionViewController.swift (hunk 4 above)
Comment (cyndichin) [reply firefox-ios/Client/Frontend/TrackingProtection/TrackingProtectionViewController.swift]: okay, we may need to do this for the error pages work as well. just wanted to make sure we're not duplicating efforts if possible. thanks!

Comment (cyndichin) [reply firefox-ios/Client/Frontend/TrackingProtection/TrackingProtectionViewController.swift]: I'll let @ih-codes review this PR more thoroughly + lean on her approval since your both doing a great job at it, feel free to ping me if you need extra eyes.

**Review by ih-codes — APPROVED** [review]
Thanks @dicarobinho ! Looking good. 😎 

I suggested only that you use `DefaultLogger` rather than `print` for errors.

I'm approving so you can merge this whenever you've fixed that last comment. 👍

File: firefox-ios/Client/Frontend/TrackingProtection/CertificatesHandler.swift [hunk 17]
```diff
@@ -32,7 +33,9 @@ class CertificatesHandler {
                 let certificate = try Certificate(derEncoded: Array(certificateData))
                 certificates.append(certificate)
             } catch {
-                print("Error decoding certificate: \(error)")
+                DefaultLogger.shared.log("\(error)",
```
Comment (ih-codes) [inline firefox-ios/Client/Frontend/TrackingProtection/CertificatesHandler.swift:36]: @dicarobinho In a future PR or task, please use your original `Error decoding certificate: \(error)"` and `"Error fetching data: \(error)"` as the log messages. Otherwise anyone seeing an error won't know where it's coming from. Thanks! 😃
